src/zenml/cli/pipeline.py

```diff
@@ -448,9 +471,122 @@ def deploy_pipeline(
+            cli_utils.declare(
+                f"[DRY-RUN] Generating deployment artifacts for '{deployment_name}'"
+            )
+
+            pipeline_instance.prepare()
+            snapshot = pipeline_instance._create_snapshot(
+                **pipeline_instance._run_args
+            )
+
+            stack = client.active_stack
+            stack.prepare_pipeline_submission(snapshot=snapshot)
+
+            if not stack.deployer:
+                cli_utils.error(
+                    "The stack does not have a deployer. Please add a "
+                    "deployer to the stack in order to perform dry-run."
+                )
+
+            # Check if the deployer supports dry-run mode
+            if not hasattr(stack.deployer, "do_dry_run_deployment"):
+                cli_utils.error(
+                    f"The '{stack.deployer.name}' deployer does not support dry-run mode.\n"
+                    f"Dry-run is only available for deployers that implement the "
+                    f"'do_dry_run_deployment' method.\n"
+                    f"Please remove the --dry-run flag to deploy normally."
+                )
+
+            from datetime import datetime, timezone
+            from uuid import uuid4
+
+            from zenml.constants import (
+                ENV_ZENML_ACTIVE_PROJECT_ID,
+                ENV_ZENML_ACTIVE_STACK_ID,
+            )
+            from zenml.models import (
+                DeploymentResponse,
+                DeploymentResponseBody,
+                DeploymentResponseMetadata,
+                DeploymentResponseResources,
+            )
+            from zenml.orchestrators.utils import get_config_environment_vars
+
+            # Create temporary deployment with resources pre-populated
+            # (avoids DB lookup when accessing deployment.snapshot)
+            now = datetime.now(timezone.utc)
+
+            # Get the deployer's ComponentResponse representation
+            deployer_response = client.zen_store.get_stack_component(
+                stack.deployer.id
+            )
+
+            temp_deployment = DeploymentResponse(
+                id=uuid4(),
+                name=deployment_name,
+                project=snapshot.project_id,
+                snapshot_id=snapshot.id,
+                deployer_id=stack.deployer.id,
+                created=now,
+                updated=now,
+                body=DeploymentResponseBody(
+                    created=now,
+                    updated=now,
+                    user_id=client.active_user.id,
+                    project_id=snapshot.project_id,
+                ),
+                metadata=DeploymentResponseMetadata(
+                    deployment_metadata={},
+                    auth_key=None,
+                ),
+                resources=DeploymentResponseResources(
+                    snapshot=snapshot,
+                    deployer=deployer_response,
+                    tags=[],
+                ),
+            )
```

Comment:
this is another thing that just feels wrong: if you have to "fake" things like deployment responses in order to get the rest of your code to behave correctly, that basically looks like you're pushing a square peg in a round hole: it's not good design.

this tells me that you need to make the base deployer class and everything else around it aware of this new "dry-run" mode